docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+}
+
+pub struct SparsePatterns {
+    paths: Vec<SparsePatternsPath>,
+    mappings: Vec<SparsePatternsMapping>,
+}
+
+pub trait OpStore {
+    ...
+    pub fn read_sparse_patterns(&self, id: &SparsePatternsIn) -> OpStoreResult<SparsePatterns> { ... }
+    pub fn write_sparse_patterns(&self, sparse_patterns: &SparsePatterns) -> OpStoreResult<SparsePatternsId> { .. }
+}
+```
+
+To support these more complex behaviors, a new `SparsePatterns` trait will be
+introduced, mainly to facilitate migration to the new data type at all the
+places where sparse patterns are currently used.
+
+```rust
+impl SparsePatterns {
+    pub fn to_matcher(&self) -> Box<dyn Matcher> {
+        ...
+    }
+
+    ...
+}
+```
+
+## Command Syntax
+
+Path rules can be specified on the CLI and in an editor via a `<type>:<path>`
+syntax. The editor will only support the new explicit syntax, whereas the CLI
+arguments will continue to support bare `<path>` arguments in a
+backwards-compatible way at least for a migration period.
+
+```rust
+pub fn type_string(type: SparsePatternsPathType) -> &'static str {
+    match type {
+        SparsePatternsPathType::DirInclusion => "include",
+        SparsePatternsPathType::DirExclusion => "exclude",
+        SparsePatternsPathType::Files => "files",
+        SparsePatternsPathType::File => "file",
+    }
+}
+```
+
+- `jj sparse set --add foo/bar` is equal to `jj sparse set --add include:foo/bar`
+- `jj sparse set --add exclude:foo/bar` adds a new `DirExclusion` type rule
+- `jj sparse set --exclude foo/bar` as a possible shorthand for the above
+- `jj sparse list` will print the explicit rules
+
+Paths will be stored in an ordered, canonical form which unambiguously describes
+the set of files to be included. Every `--add` command will append to the end of
+this list before the patterns are canonicalized. Whether a file is included is
+determined by the first matching rule in reverse order.
+
+For example:
+
+```txt
+include:foo
+exclude:foo/bar
+include:foo/bar/baz
+exclude:foo/bar/baz/qux
+```
+
+Produces rule set which includes "foo/file.txt", excludes "foo/bar/file.txt",
+includes "foo/bar/baz/file.txt", and excludes "foo/bar/baz/qux/file.txt". If the
+rules are subtly re-ordered, they become canonicalized to a smaller, but
+functionally equivalent form:
+
+```txt
+# Before
+include:foo
+exclude:foo/bar/baz/qux
+include:foo/bar/baz
+exclude:foo/bar
+
+# Canonicalized
+include:foo
+exclude:foo/bar
+```
+
+## Working Copy Map
+
+WARNING: This section is intentionally lacking, more research is needed.
+
+All Sparse Patterns v2 will come equipped with a default no-op mapping.
+
+```rust
+vec![WorkingCopyMap {
+    src_path: RepoPathBuf::root(),
+    dst_path: RepoPathBuf::root(),
+}]
+```
+
+`SparsePatterns` will provide an interface to map client paths into repo paths,
+and vice versa. The WorkingCopy trait will apply this mapping to all snapshot
+and checkout operations, and and jj commands which accept relative paths will
```

Comment:
This is mostly a nudge to consider it, even if we discard the idea. Also a good use-case to mention P4's client maps, which seem similar.

Reply:
> That seems infeasible (or at least undesirable), especially if the patterns are version controlled, unless the contents of the path is also somehow version controlled.

I don't expect the path to be version controlled, rather it is a client-side option for CI which is only interested in particular files, like iOS/Android builds or just testdata (fuzzer-corpora, golden tests, etc) . So a configuration option per user or per repo.  

> I think it's undesirable even if it's not because it seems unclear what should happen if you edit the patterns. Would you then run something like `jj patterns refresh`?

I'd wish that we had a client-side watcher for such a file or manually refreshing would also work.

Reply:
> I added a like to P4 client maps here for completeness.

Thanks. 

> I also think having this be configuration-controlled could lead to a lot of badness.

Could a safe subset of it exist at any point or is this another thing I'm currently not thinking through correctly? I'm fine to drop configuration as I had two authorative No's.

Reply:
The issue I see with it being configuration controlled is that if you ever run `jj sparse <anything>`, it'll transparently succeed but be automatically undone by the next command.

I guess we could identify if the config exists and then error out early on any commands that would attempt to modify the WorkingCopyPatterns?  That still feels like a weird situation to support.  Also, if the path to the patterns is itself version controlled, what happens when that file designates patterns that exclude itself?

Reply:
> The issue I see with it being configuration controlled is that if you ever run `jj sparse <anything>`, it'll transparently succeed but be automatically undone by the next command.

Yes that seems to be problematic for human users but for automation which only expects _their_ subset of files available it shouldn't matter. The goal of having this configuration option is to have a subset of what Google and Meta do with their respective VCS clients w/o it being available.

> I guess we could identify if the config exists and then error out early on any commands that would attempt to modify the WorkingCopyPatterns? That still feels like a weird situation to support.

Yes, with that it's definitely a no-go for a human but in my opinion this does not rule automation out yet.

> Also, if the path to the patterns is itself version controlled, what happens when that file designates patterns that exclude itself?

As previously said I wouldn't allow the file to be in the repo, I'd just ship it to the respective machines in the respective configuration dir.

Reply:
I added an appendix section for the configuration idea.